Fix upload/download issue of objects containing a "+" in their key

JBWatenbergScality opened the pull request.

JBWatenbergScality committed:
Fix upload/download issue of objects containing a "+" in their key

bert-e commented:
# Hello jbwatenbergscality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/zenko-ui/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ZKUI-383 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `2.1.4`


Please check the `Fix Version/s` of ZKUI-383, or the target
branch of this pull request.

JBWatenbergScality commented:
/approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @monpote

* @chengyanjin

* @jbwatenbergscality










*The following options are set:* **approve**

JBWatenbergScality committed:
fix: upload, download and listing of objects with "+" in their keys
When uploading an object containing several + in their keys, nginx considers the "+" as being a "valid" url charcter, then it is not re-encoding it when proxy_pass the url.

This ends up in signature validation issues on cloudserver side, because cloud server receives the request with a plain "+" character instead of its urlencode counterpart "%2B".
To solve the issue, we create a new variable named "urlencore_proxy_uri" where we urlencode the "+" and give that URL to the proxy_pass, this way nginx is not re-encoding it.

We had to introduce lua to do this operation via a set_by_lua_block because a simple map module in nginx is not replacing ALL the occurences but only the last match.

Ref: ZKUI-317

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/zenko-ui/commit/1cf657b21c1acd423c7825b1a214b5a0a8b23320) did not succeed in branch bugfix/ZKUI-383.




*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.1`



The following branches will **NOT** be impacted:

* `development/1.4`
* `development/1.5`
* `development/1.6`
* `development/2.0`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.1`



The following branches have **NOT** changed:

* `development/1.4`
* `development/1.5`
* `development/1.6`
* `development/2.0`



Please check the status of the associated issue ZKUI-383.

Goodbye jbwatenbergscality.
